File: ios/Approach/Sources/StatisticsLibrary/Protocols/CountingStatistic.swift
import Charts
import ModelsLibrary

public protocol CountingStatistic: Statistic {
	var count: Int { get set }
}

extension CountingStatistic {
	public static var supportsAggregation: Bool {
		true
	}

	public static var supportsWidgets: Bool {
		true
	}

	public mutating func aggregate(with: Statistic) {
		guard let with = with as? Self else { return }
		self.count += with.count
	}

	public var formattedValue: String {
		String(count)
	}

	public var formattedValueDescription: String? {
		nil
	}

	public var isEmpty: Bool {
		// `count` is defined on `CountingStatistic` and we are defining `isEmpty` here, so we can't _use_ `isEmpty`
		// swiftlint:disable:next empty_count
		count == 0
	}
}

File: android/core/statistics/src/main/java/ca/josephroque/bowlingcompanion/core/statistics/interfaces/CountingStatistic.kt
package ca.josephroque.bowlingcompanion.core.statistics.interfaces

import ca.josephroque.bowlingcompanion.core.statistics.Statistic

interface CountingStatistic : Statistic {
	var count: Int

	override val supportsAggregation: Boolean
		get() = true

	override val supportsWidgets: Boolean
		get() = true

	override val formattedValue: String
		get() = count.toString()

	override val formattedValueDescription: String?
		get() = null

	override val isEmpty: Boolean
		get() = count == 0

	override fun aggregateWithStatistic(statistic: Statistic) {
		if (statistic is CountingStatistic) {
			count += statistic.count
		}
	}
}
